### PR TITLE
AudioMixer resampler for multi-channel output

### PR DESCRIPTION
**Pull request by haochuan1982**
Correct AudioMixer's channel setting for multi-channel
Update AudioMixer mono playback with resampler behavior

Tracked-On: OAM-111211

### CODE REVIEW

File: aosp_diff/celadon_ivi/frameworks/av/media/libaudioprocessing/0004-AudioMixer-resampler-for-multi-channel-output.patch [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
++++ b/media/libaudioprocessing/AudioMixerBase.cpp
+@@ -490,7 +490,17 @@ bool AudioMixerBase::TrackBase::setResampler(uint32_t trackSampleRate, uint32_t
+ 
+                 // TODO: Remove MONO_HACK. Resampler sees #channels after the downmixer
+                 // but if none exists, it is the channel count (1 for mono).
+-                const int resamplerChannelCount = getOutputChannelCount();
++                //const int resamplerChannelCount = getOutputChannelCount();
++                int reqChannel = 0;
++                uint32_t mixerChannelCount = getMixerChannelCount();
++                uint32_t inChannelCount = getOutputChannelCount();
```
Comment (libinyang) [inline aosp_diff/celadon_ivi/frameworks/av/media/libaudioprocessing/0004-AudioMixer-resampler-for-multi-channel-output.patch:22]: outChannelCount?

**Comment by xinxin-wan** [comment]
LGTM

**Review by gkdeepa — APPROVED** [review]
LGTM

File: aosp_diff/celadon_ivi/frameworks/av/media/libaudioprocessing/0004-AudioMixer-resampler-for-multi-channel-output.patch [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+index a56fc9060a..f8c02814e6 100644
+--- a/media/libaudioprocessing/AudioMixerBase.cpp
++++ b/media/libaudioprocessing/AudioMixerBase.cpp
+@@ -490,7 +490,29 @@ bool AudioMixerBase::TrackBase::setResampler(uint32_t trackSampleRate, uint32_t
+ 
+                 // TODO: Remove MONO_HACK. Resampler sees #channels after the downmixer
+                 // but if none exists, it is the channel count (1 for mono).
+-                const int resamplerChannelCount = getOutputChannelCount();
++                //const int resamplerChannelCount = getOutputChannelCount();
++                /* Martin, Chen: this is maybe android native bug.
```
Comment (JeevakaPrabu) [inline aosp_diff/celadon_ivi/frameworks/av/media/libaudioprocessing/0004-AudioMixer-resampler-for-multi-channel-output.patch:22]: Please remove "Martin, Chen: "

Comment (JeevakaPrabu) [reply aosp_diff/celadon_ivi/frameworks/av/media/libaudioprocessing/0004-AudioMixer-resampler-for-multi-channel-output.patch:22]: Detailed commit message should have details about the issue and fix

**Review by JeevakaPrabu — CHANGES_REQUESTED** [review]

**Review by JeevakaPrabu — APPROVED** [review]

**Review by JeevakaPrabu — APPROVED** [review]

**Pull request merged**